src/server/db/migrations/2020-11-13_0.3.0_to_0.4.0/migrate_0.3.0_to_0.4.0.js

```diff
@@ -0,0 +1,414 @@
+    };
+
+    return gpii.dbRequest.configureStep(details, options);
+};
+
+/**
+ * Log how many documents were updated
+ *
+ * @param {String} responseString - Response from the database (ignored)
+ * @param {Object} options - Object containing the set of documents:
+ * @param {Array} options.updatedDocs - The documents to update
+ *
+ * @return {Number} - the number of documents updated.
+ */
+sjrk.storyUpdate.logUpdateDB = function (responseString, options) {
+    fluid.log("-- -- --");
+    fluid.log("Updated ", options.updatedDocs.length, " of ", options.totalNumOfDocs, " story documents.");
+    fluid.log("-- -- --");
+    return options.updatedDocs.length;
+};
+
+/**
+ * Create the step that retrieves stories from the database
+ *
+ * @param {Object} options - The documents to be verified:
+ * @param {Array} options.updatedDocs - The documents to verify
+ *
+ * @return {Promise} - The promise that resolves the verificatioon
+ */
+sjrk.storyUpdate.verifyStories = function (options) {
+    var details = {
+        requestUrl: options.storiesUrl,
+        requestErrMsg: "Error retrieving stories from the database: ",
+        responseDataHandler: sjrk.storyUpdate.verifyStoriesData,
+        responseErrMsg: "Error retrieving stories from database: "
+    };
+    return gpii.dbRequest.configureStep(details, options);
+};
+
+/**
+ * Check all database documents for errors related to the data update
+ *
+ * @param {String} responseString - the response from the request to get all documents
+ * @param {Object} options - Where to store any potentially-discovered invalid documents
+ *
+ * @return {Array} - The invalid documents found during verification
+ */
+sjrk.storyUpdate.verifyStoriesData = function (responseString, options) {
+    var allDocs = JSON.parse(responseString);
+
+    if (allDocs.total_rows !== options.totalNumOfDocs) {
+        throw "Number of records retrieved for verification does not match number of records retrieved for updating";
+    }
+
+    var invalidDocs = [];
+
+    if (allDocs.rows) {
+        fluid.each(allDocs.rows, function (aRow) {
+            var aDoc = aRow.doc;
+            // To filter out the "_design/views" doc that doesn't have the "schemaVersion" field
+            if (aDoc.type === "story") {
+
+                // assume all stories were successfully migrated
+                // if an error is discovered, the cause will be tracked
+                var invalidStoryReasons = [];
+
+                if (!aDoc.schemaVersion || aDoc.schemaVersion < sjrk.storyUpdate.newSchemaVersion) {
+                    invalidStoryReasons.push("Story schema version is not up to date");
+                }
+
+                if (!aDoc.value.id || aDoc.value.id !== aDoc._id) {
+                    invalidStoryReasons.push("Story ID and internal database ID do not match");
+                }
+
+                var requiredStoryValues = [
+                    "published",
```

Comment:
I wonder rather than checking all non-existent keys now exist, shall their values be checked as well? For example:

1. published === true;
2. three order values are set to correct values.

Reply:
for the `published` flag, there may be unpublished stories in the future, so I opted to keep it a bit more general. In this case, since story data is not expected to have that value beforehand, checking that value is reasonable.

I'll add checking for the order values, since those are predictable.

Reply:
Well, @BlueSlug, i wonder how future is "in the future"? If the first part of the migration script is looping thru every single doc to set the published flag to `true`, shouldn't the 2nd verification step to make sure this action has happened correctly to all docs?

If at the point this migration script is executed, there aren't any unpublished stories, I think we need to make sure at that point of time, all updates are correct.

My point is migration scripts that run and only run once at a certain point of time are different from the regular tool scripts that consider the future support.

Reply:
That's a good point! I'll check the value in this case as well, then.